protocols/autonatv2/src/client/behaviour.rs

```diff
@@ -0,0 +1,294 @@
+                self.handle_no_connection(peer_id, connection_id);
+            }
+            _ => {}
+        }
+    }
+
+    fn on_connection_handler_event(
+        &mut self,
+        peer_id: PeerId,
+        connection_id: ConnectionId,
+        event: <Handler as ConnectionHandler>::ToBehaviour,
+    ) {
+        self.peers_to_handlers
+            .entry(peer_id)
+            .or_insert(connection_id);
+        match event {
+            Either::Right(Ok(nonce)) => {
+                if self.pending_nonces.remove(&nonce) {
+                    tracing::trace!("Received pending nonce from {peer_id:?}");
+                } else {
+                    tracing::warn!("Received unexpected nonce from {peer_id:?}, this means that another node tried to be reachable on an address this node is reachable on.");
+                }
+            }
+            Either::Right(Err(err)) => {
+                tracing::debug!("Dial back failed: {:?}", err);
+            }
+            Either::Left(dial_request::ToBehaviour::PeerHasServerSupport) => {
+                if !self.known_servers.contains(&peer_id) {
+                    self.known_servers.push(peer_id);
+                }
+            }
+            Either::Left(dial_request::ToBehaviour::TestCompleted(Ok(TestEnd {
+                dial_request: DialRequest { nonce, addrs },
+                suspicious_addr,
+                reachable_addr,
+            }))) => {
+                if self.pending_nonces.remove(&nonce) {
+                    tracing::debug!(
+                        "server reported reachbility, but didn't actually reached this node."
+                    );
+                    return;
+                }
+                if !suspicious_addr.is_empty() {
+                    tracing::trace!(
+                        "server reported suspicious addresses: {:?}",
+                        suspicious_addr
+                    );
+                }
+                self.pending_events.extend(
+                    addrs
+                        .into_iter()
+                        .take_while(|addr| addr != &reachable_addr)
+                        .map(ToSwarm::ExternalAddrExpired),
+                );
+                self.pending_events
+                    .push_back(ToSwarm::ExternalAddrConfirmed(reachable_addr));
+            }
+            Either::Left(dial_request::ToBehaviour::TestCompleted(
+                Err(dial_request::Error::UnableToConnectOnSelectedAddress { addr: Some(addr) })
+            ))
+            | Either::Left(dial_request::ToBehaviour::TestCompleted(
+                Err(dial_request::Error::FailureDuringDialBack { addr: Some(addr) })
+            )) => {
+                self.pending_events
+                    .push_back(ToSwarm::ExternalAddrExpired(addr));
+            }
+            Either::Left(dial_request::ToBehaviour::TestCompleted(Err(err))) => {
+                tracing::debug!("Test failed: {:?}", err);
+            }
```

Comment:
Is there any state we need to clean up upon failures in order to not cause a memory leak?

Reply:
I did that clean up.